src/diagnostics/MOM_harmonic_analysis.F90

```diff
@@ -0,0 +1,411 @@
+                                        phase0      !< The phase of a tidal constituent at time 0 [rad]
+  integer,               intent(in)  :: nc          !< The number of tidal constituents in use
+  character(len=16),     intent(in)  :: const_name(MAX_CONSTITUENTS) !< The name of each constituent
+  type(HA_CS),           intent(out) :: CS          !< Control structure of the MOM_harmonic_analysis module
+
+  ! Local variables
+  type(HA_type) :: ha1                              !< A temporary, null field used for initializing CS%list
+  real :: HA_start_time                             !< Start time of harmonic analysis [day]
+  real :: Time_unit                                 !< The time unit for CS%time_end [s]
+  character(len=40)  :: mdl="MOM_harmonic_analysis" !< This module's name
+  character(len=255) :: mesg
+  integer :: year, month, day, hour, minute, second
+
+  integer :: unit, io_status
+  integer :: date_init(6)=0                         !< The start date of the whole simulation.
+  character(len=16) :: calendar = 'julian'          !< The name of the calendar type.
+  integer :: years=0, months=0, days=0              !< These may determine the segment run
+  integer :: hours=0, minutes=0, seconds=0          !! length, if read from a namelist
+
+  namelist /ocean_solo_nml/ date_init, calendar, months, days, hours, minutes, seconds
+
+  ! Populate some parameters of the control structure
+  CS%time_ref   =  time_ref
+  CS%freq       =  freq
+  CS%phase0     =  phase0
+  CS%nc         =  nc
+  CS%const_name =  const_name
+  CS%length     =  0
+
+  ! Initialize CS%list
+  allocate(CS%list)
+  CS%list%this  =  ha1
+  nullify(CS%list%next)
+
+  ! Determine CS%time_end by end time of the run segment
+  if (file_exists('input.nml')) then
+    call open_ASCII_file(unit, 'input.nml', action=READONLY_FILE)
+    read(unit, ocean_solo_nml, iostat=io_status)
+    call close_file(unit)
+  endif
+
+  if (years+months+days+hours+minutes+seconds > 0) then
+    CS%time_end = increment_date(Time, years, months, days, hours, minutes, seconds)
+    call MOM_mesg('HAmod: end time of harmonic analysis determined from ocean_solo_nml.', 2)
+  else
+    call get_param(param_file, mdl, "TIMEUNIT", Time_unit, &
+                   "The time unit for DAYMAX, ENERGYSAVEDAYS, and RESTINT.", &
+                   units="s", default=86400.0)
+    call get_param(param_file, mdl, "DAYMAX", CS%time_end, &
+                   "The final time of the whole simulation, in units of "//&
+                   "TIMEUNIT seconds.  This also sets the potential end "//&
+                   "time of the present run segment if the end time is "//&
+                   "not set via ocean_solo_nml in input.nml.", &
+                   timeunit=Time_unit, fail_if_missing=.true.)
+  endif
+
+  ! Determine CS%time_start
+  call get_param(param_file, mdl, "HA_start_time", HA_start_time, &
```

Comment:
Because our param_file parser is case sensitive, we have adopted the (perhaps poorly documented) convention that we use all upper case for parameters, so in this case "HA_START_TIME" would be preferable over "HA_start_time".